Integrated JS debugger (#115)
Tecnativa/doodba-copier-template: docs/daily-usage.md

To browse Odoo go to `http://localhost:${ODOO_MAJOR}069` (i.e. for Odoo 11.0 this would
be `http://localhost:11069`).

**Note:** For more information, check

> [Visual Studio Code Integration](#visual-studio-code-integration)

Author: This note should go above, in this paragraph: https://github.com/Tecnativa/doodba-copier-template/pull/115/files#diff-db5b9c6fa770aa40cf2f944e5129d29dR71-R72

**⚠️ DO NOT USE IT IN PRODUCTION ENVIRONMENTS ⚠️** (I had to say it).

> **Note:** For more information, check [Debugging Python code](#debugging-python-code)

Author: wdb is a different debugger, completely unrelated to VSCode, so this is more exact:

```suggestion
**Note:** Remember that you can instead [use the integrated debugger in Visual Studio Code](#debugging-python-code)
```

</details>

## Visual Studio Code Integration

> This project will integrate with it [Visual Studio Code](https://code.visualstudio.com/)

Author: ```suggestion
Generated projects have out-of-the-box integration with [Visual Studio Code](https://code.visualstudio.com/)
```

to provide you with a better and easier-to-use experience through:

- [Suggested extensions](/.vscode/extensions.json)
- Launch configurations (See section on [tasks file](/tasks.py.jinja))
- Workspace definition (See section on [tasks file](/tasks.py.jinja))
- Integration with invoke (See section on [tasks file](/tasks.py.jinja))
- Debugger integration (See `tasks` section on [tasks file](/tasks.py.jinja) and

>   sections below)

Author: Remove links, I think those can be confusing.

> 

Author: The notes you added twice below should be here I think:

```suggestion

**Note**: As the configurations are
defined per project and depending on the Odoo addons you are working on, you might
need to update your configuration frequently with `invoke write-code-workspace-file`
(done automatically when you invoke other tasks).
```

> ### Debugging Python code

Author: I miss the section about debugging python + JS code, which I think is the most awesome new feature.

Reviewer: Pls check if that's what you had in mind

In order to debug Python code, you need to:

1. Install suggested extensions, as mentioned above

> 2. Run the project with the provided `invoke` commands. (This will automatically
>    generate the VSCode configurations for you) **Note**: As the configurations are
>    defined per project and depending on the Odoo addons you are working on, you might
>    need to update your configuration frequently with `invoke write-code-workspace-file`
>    (done automatically when you do `invoke start`, for example).

Author: ```suggestion
2. Run `invoke develop`.
```

3. Go to the "Run" tab in VSCode, select "Start Odoo and debug Python", and click "Run".
   The program will stop in any breakpoint you have and you will be able to analyze your
   Python environment.

### Debugging JS code

In order to debug JavaScript code, you need to:

1. Install suggested extensions, as mentioned above

> 2. Run the project with the provided `invoke` commands. (This will automatically
>    generate the VSCode configurations for you) **Note**: As the configurations are
>    defined per project and depending on the Odoo addons you are working on, you might
>    need to update your configuration frequently with `invoke write-code-workspace-file`
>    (done automatically when you do `invoke start`, for example).

Author: ```suggestion
2. Run `invoke develop`.
```

3. Go to the "Run" tab in VSCode, select "Start Odoo and debug JS in [Firefox/Chrome]",
   and click "Run". The program will stop in any breakpoint you have and you will be
   able to analyze your JS environment.

> 

Author: ```suggestion

If you face some problems while debugging in Firefox or Chrome, see our [FAQ](faq.md) or the corresponding extensions documentation before opening an issue here.
```

`invoke write-code-workspace-file` (done automatically when you invoke other tasks).

### Debugging Python code

In order to debug Python code, you need to:

1. Install suggested extensions, as mentioned above
1. Run `invoke develop`
1. Go to the "Run" tab in VSCode, select "Start Odoo and debug Python", and click "Run".
   The program will stop in any breakpoint you have and you will be able to analyze your
   Python environment.

### Debugging JS code

In order to debug JavaScript code, you need to:

1. Install suggested extensions, as mentioned above
1. Run `invoke develop`
1. Go to the "Run" tab in VSCode, select "Start Odoo and debug JS in [Firefox/Chrome]",
   and click "Run". The program will stop in any breakpoint you have and you will be
   able to analyze your JS environment.

If you face some problems while debugging in Firefox or Chrome, see our [FAQ](faq.md) or
the corresponding extensions documentation before opening an issue here.

### Debugging Python + JS code 🤯

In order to debug Python and JavaScript code at the same time, you need to:

1. Install suggested extensions, as mentioned above

> 1. Run the project with the provided `invoke` commands. (This will automatically
>    generate the VSCode configurations for you) **Note**: As the configurations are
>    defined per project and depending on the Odoo addons you are working on, you might
>    need to update your configuration frequently with `invoke write-code-workspace-file`
>    (done automatically when you do `invoke start`, for example).

Author: You don't really need to do this, because the pre launch task does this for you:

```suggestion
```

1. Install suggested extensions, as mentioned above
2. Run the project with the provided `invoke` commands. (This will automatically
   generate the VSCode configurations for you) **Note**: As the configurations are
   defined per project and depending on the Odoo addons you are working on, you might
   need to update your configuration frequently with `invoke write-code-workspace-file`
   (done automatically when you do `invoke start`, for example).
3. Go to the "Run" tab in VSCode, select "Start Odoo and debug Python", and click "Run".
   The program will stop in any breakpoint you have and you will be able to analyze your
   Python environment.

### Debugging JS code

In order to debug JavaScript code, you need to:

1. Install suggested extensions, as mentioned above
2. Run the project with the provided `invoke` commands. (This will automatically
   generate the VSCode configurations for you) **Note**: As the configurations are
   defined per project and depending on the Odoo addons you are working on, you might
   need to update your configuration frequently with `invoke write-code-workspace-file`
   (done automatically when you do `invoke start`, for example).

> 3. Go to the "Run" tab in VSCode, select "Start Odoo and debug JS in [Firefox/Chrome]",

Author: Always use 1 please, to avoid confusion if some points are added or removed in between:

```suggestion
1. Go to the "Run" tab in VSCode, select "Start Odoo and debug JS in [Firefox/Chrome]",
```

   and click "Run". The program will stop in any breakpoint you have and you will be
   able to analyze your JS environment.

If you face some problems while debugging in Firefox or Chrome, see our [FAQ](faq.md) or
the corresponding extensions documentation before opening an issue here.

### Debugging Python + JS code 🤯

In order to debug Python and JavaScript code at the same time, you need to:

1. Install suggested extensions, as mentioned above
2. Run the project with the provided `invoke` commands. (This will automatically
   generate the VSCode configurations for you) **Note**: As the configurations are
   defined per project and depending on the Odoo addons you are working on, you might
   need to update your configuration frequently with `invoke write-code-workspace-file`
   (done automatically when you do `invoke start`, for example).

> 3. Go to the "Run" tab in VSCode, select "Start Odoo and debug Python + JS in

Author: ```suggestion
1. Go to the "Run" tab in VSCode, select "Start Odoo and debug Python + JS in
```

## Visual Studio Code Integration

Generated projects have out-of-the-box integration with
[Visual Studio Code](https://code.visualstudio.com/) to provide you with a better and
easier-to-use experience through:

> - Suggested extensions
> - Launch configurations (See section on `tasks` file)
> - Workspace definition (See section on `tasks` file)
> - Integration with invoke (See section on `tasks` file)
> - Debugger integration (See section on `tasks` file and sections below)

Author: ```suggestion
- Suggested extensions
- Launch configurations.
- Workspace definition.
- Integration with invoke.
- Debugger integration.
```

   and click "Run". The program will stop in any breakpoint you have and you will be
   able to analyze your JS environment.

If you face some problems while debugging in Firefox or Chrome, see our [FAQ](faq.md) or
the corresponding extensions documentation before opening an issue here.

### Debugging Python + JS code 🤯

In order to debug Python and JavaScript code at the same time, you need to:

1. Install suggested extensions, as mentioned above

> 2. Run the project with the provided `invoke` commands. (This will automatically
>    generate the VSCode configurations for you) **Note**: As the configurations are
>    defined per project and depending on the Odoo addons you are working on, you might
>    need to update your configuration frequently with `invoke write-code-workspace-file`
>    (done automatically when you do `invoke start`, for example).

Author: The launchers that start with "Start Odoo" already do this for you, so you can remove this step:

```suggestion
```

3. Go to the "Run" tab in VSCode, select "Start Odoo and debug Python + JS in
   [Firefox/Chrome]", and click "Run". The program will stop in any breakpoint you have
   and you will be able to analyze your Python and JS environment. You will just need to
   switch between both environments running at the same time in the VSCode debugger.

>    ![](res/vscode_py_js_debug.gif)

Author: ```suggestion
   ![Switching debugger](res/vscode_py_js_debug.gif)
```

> 
> If you face some problems while debugging in Firefox or Chrome, see our [FAQ](faq.md) or
> the corresponding extensions documentation before opening an issue here.

Author: DRY:

```suggestion
```
